01-data-structures/03-linked-lists/linked-lists-answers.txt

```diff
@@ -0,0 +1,29 @@
+1. What is Spatial Locality and why does it benefit performance?
+
+Spatial locality is the concept that you are able to find data quicker if their addresses or memory locations are close by. It benefits performance because it reduces the number of lookups.
+
+2. Compare the performance of an Array to a Linked List using the Benchmark module.
+
+Compare the time it takes to create a 10,000 item Array to appending 10,000 items to a Linked List.
+
+Push 10,000 Items
+              user     system      total        real
+array        0.000000   0.000000   0.000000 (  0.001615)
+linked list  1.180000   0.000000   1.180000 (  1.189786)
+
+
+Compare the time it takes to access the 5000th element of the Array and the 5000th Node in the Linked List.
+
+Access 5,000th Element/Node
+              user     system      total        real
+array        0.000000   0.000000   0.000000 (  0.000002)
+linked list  0.000000   0.000000   0.000000 (  0.000408)
+
+Compare the time it takes to remove the 5000th element from the Array to removing the 5000th Node in the Linked List. In the Array, the 5001st item becomes the 5000th, and so on.
+
+Remove 5,000th Element/Node
+              user     system      total        real
+array        0.000000   0.000000   0.000000 (  0.000004)
+linked list  0.010000   0.000000   0.010000 (  0.001126)
+
+Across all cases, the performance of an array is better than a linked list largely due to spatial locality. With arrays, because elements are positioned sequentially, we can access any element with an index. On the other hand, with a linked list, we need to traverse the list node by node until we find what we are looking for. Once we moved past a node, we cannot move backwards. A node only has information about what the next node is but not the previous node.
```

Comment:
It is also worth noting that it is hardly fair to pit a professionally developed dynamic array written in C to a student-developed linked list written in Ruby =P